### PR TITLE
Feature/helm chart use gratia image

### PR DESCRIPTION
**Pull request by mwestphall**
- Update values.yaml with configuration options for a Gratia output container
- Update cronjob.yaml to optionally run the Gratia output container rather than the APEL output container
- Update cronjob.yaml to run the processor initContainer via a custom docker image rather than by pulling the repo

The Docker images that are run for both the processor initContainer and the Gratia output container are currently being manually built and pushed into OSG Harbor. We will need to design a CI workflow before these changes to the helm chart can be used in production.

### CODE REVIEW

**Comment by brianhlin** [comment]
@rptaylor once we sort out this PR, I'd like to nail down details of the Helm chart build pipeline + release. What do you think about the idea of using `git filter-repo` and taking the helm chart bits from this repo and dumping them into https://github.com/osg-htc/helm-charts under `supported/iris-hep/kuantifier`? We'd add you as a code owner to give you access to that area

File: chart/templates/cronjob.yaml
Comment (rptaylor) [inline chart/templates/cronjob.yaml:32]: Are there any cases where it would need to run as root? It's completely unprivileged.

File: chart/README.md
Comment (rptaylor) [inline chart/README.md:16]: Missing link? Though TBH how to use Helm is a bit out of scope, but happy to add if it helps people.

File: chart/README.md
Comment (rptaylor) [inline chart/README.md:56]: The code only cares about pods, "pilot job" is a bit specific to certain experiment workflows.

File: chart/templates/cronjob.yaml
Comment (mwestphall) [reply chart/templates/cronjob.yaml:32]: @rptaylor On second inspection this may be an issue pertaining to cluster configuration as well. I'm also receiving an error related to runAsNoonRoot when running the production kapel helm chart in CHTC's kubernetes cluster, since the `library/python:3` image runs as root by default.
```
Error: container has runAsNonRoot and image will run as root (pod: "kapel-test-job", container: processor)
```
The [documentation that I can find](https://kubernetes.io/docs/concepts/security/pod-security-standards/#restricted) pertaining to `securityContext.runAsNonRoot` indicates that the flag just checks whether a pod's containers runs as root and throws an error if they do. Should we add additional configuration (such as [`runAsUser: UID > 0`](https://kubernetes.io/docs/tasks/configure-pod-container/security-context/))?

Comment (rptaylor) [reply chart/templates/cronjob.yaml:32]: Ah, unfortunately our cluster is a bit old and still using PSP, one of the potential pitfalls of PSP is that it is mutating and can change the default runAsUser. For us it runs as UID 10000 by default. But we should make it explicit in the cronJob spec instead of relying on the default of the image, or any changes that extraneous PSP might apply.

I still think having a fixed value `runAsNonRoot: true` is a good safeguard, and yes let's add variables for runAsUser and runAsGroup. The default UID, GID could be 10000, it's just arbitrary.

Sometimes Openshift needs to configure runAs* so this will help for compatibility too.

Comment (mwestphall) [reply chart/templates/cronjob.yaml]: Just updated the helm chart to return the fixed `runAsNonRoot: true` and make runAsUser/Group configurable. We had a couple issues getting the gratia output image to run with a non-root user since the gratia rpms install with root ownership by default, we were able to selectively chmod things in the Dockerfile to get around that though.

File: chart/templates/gratia-output-config.yaml
Comment (rptaylor) [inline chart/templates/gratia-output-config.yaml:1]: This whole file should be wrapped in "if .Values.outputFormat = gratia" don't you think?

I suppose the same is true for the ssmsend-config and ssmsend-secret templates, if the format is ssmsend.

File: chart/values.yaml
Comment (rptaylor) [inline chart/values.yaml:84]: This comment is a bit misleading, the default here is actually "latest" but if a user manually sets `image_tag: ""` then .Chart.AppVersion would take effect in the cronjob template. 
Using latest is not recommended for production, so I would suggest a similar pattern to the ssmsend image_tag. Users can always set "latest" in their values if they want. Up to you though.

File: chart/templates/cronjob.yaml
Comment (rptaylor) [reply chart/templates/cronjob.yaml]: Wouldn't the gratia RPM installation happen at the container image build time? But anyway , okay good that it is all working unprivileged.

File: chart/README.md
Comment (rptaylor) [reply chart/README.md]: I updated a bit more, in principle it does not even matter whether the pods are from jobs or potentially other controllers, but I have it in the back of my mind to document the requirement that pods need to be recorded on the cluster in a completed state. The fact that in practice this seems to work well for jobs but not other pod controllers has to do with specifics of Kubernetes' implementation of different controllers and/or how KSM works.

**Comment by rptaylor** [comment]
Thanks @mwestphall for the updates, it's looking good, just added a few more comments.
Can you also point me at the Containerfile/CI being used for building the processor container?
`image_repository: "hub.opensciencegrid.org/osgpreview/kapel-processor"`

Also not sure if this is just personal preference, but for https://github.com/rptaylor/kapel/pull/53/files#diff-25d035a270f9563fd3aa0f9f0c430a034e429456f46a9aeadfe669b1b6a34f33L58  , to me it makes more sense to define the command in k8s YAML rather than relying on the default entrypoint command used in an image build; it's more visible and explicit, and in case debugging is needed it gives you an easier starting point. Not sure what the most typical/recommended practice is though. What do you think?

Thanks!

**Comment by mwestphall** [comment]
> Thanks @mwestphall for the updates, it's looking good, just added a few more comments. Can you also point me at the Containerfile/CI being used for building the processor container? `image_repository: "hub.opensciencegrid.org/osgpreview/kapel-processor"`
> 
> Also not sure if this is just personal preference, but for https://github.com/rptaylor/kapel/pull/53/files#diff-25d035a270f9563fd3aa0f9f0c430a034e429456f46a9aeadfe669b1b6a34f33L58 , to me it makes more sense to define the command in k8s YAML rather than relying on the default entrypoint command used in an image build; it's more visible and explicit, and in case debugging is needed it gives you an easier starting point. Not sure what the most typical/recommended practice is though. What do you think?
> 
> Thanks!

We don't have CI set up for the processor docker image yet, just a [Dockerfile](https://github.com/mwestphall/kapel/blob/feature/gratia-output-helm-chart/Dockerfile) and [docker-compose.yaml](https://github.com/mwestphall/kapel/blob/feature/gratia-output-helm-chart/docker-compose.yml) in a branch on my fork of the repo that's ahead of this PR by a few commits. The Dockerfile should be largely a transcription of the install/run command from the current helm chart. Then with regards to specifying an explicit command in the helm chart, I think that makes sense from a self-documentation perspective.

File: chart/values.yaml
Comment (mwestphall) [reply chart/values.yaml:84]: Thanks for catching this, I've updated the `image_tag` value to `""` in the values.yaml

File: chart/templates/gratia-output-config.yaml
Comment (mwestphall) [reply chart/templates/gratia-output-config.yaml:2]: That makes sense, I've updated all 3 files.

File: README.md
Comment (rptaylor) [inline README.md:30]: This section is meant to document the external requirements, i.e. things you need before you start configuring KAPEL. Values like this one for configuring KAPEL itself are documented in the chart docs.

File: chart/templates/cronjob.yaml
Comment (rptaylor) [inline chart/templates/cronjob.yaml:34]: The documentation says this should be `.Values.user` rather than serviceUser, let's go with the former?

Comment (rptaylor) [reply chart/templates/cronjob.yaml:34]: The values.yaml also has serviceUser.

**Comment by rptaylor** [comment]
Thanks a lot for all your work on this @mwestphall ! I just noticed one other inconsistency, other than that I can go ahead and merge it tomorrow if everything is ready.

**Comment by mwestphall** [comment]
@rptaylor Thanks for all the reviews! It should be good to merge in now. One last thought I had is that it might not be the most convenient to version all 3 of the processor initContainer image, gratia output image, and ssmsend image in lock-step with the chart appVersion, but that should be fine at least for this next release.

**Comment by mwestphall** [comment]
@rptaylor Is there anything else you'd like to address before we merge these changes in? We will need you to submit an approving review before we can apply the merge.

**Review by rptaylor — APPROVED** [review]

**Pull request merged**
